test/jdk/jdk/incubator/vector/AbstractVectorLoadStoreTest.java

```diff
@@ -32,28 +35,107 @@
+            withToString("int[i * 5]", (int s) -> {
+                return fill(s * SHUFFLE_BUFFER_REPS,
+                        i -> (int)(i * 5));
+            }),
+            withToString("int[i + 1]", (int s) -> {
+                return fill(s * SHUFFLE_BUFFER_REPS,
+                        i -> (((int)(i + 1) == 0) ? 1 : (int)(i + 1)));
+            })
+    );
+
+    static final List<IntFunction<Integer>> SHUFFLE_INDEX_GENERATORS = List.of(
+            withToString("-1", (int l) -> {
+                return -1;
+            }),
+            withToString("l", (int l) -> {
+                return l;
+            }),
+            withToString("l - 1", (int l) -> {
+                return l - 1;
+            }),
+            withToString("l + 1", (int l) -> {
+                return l + 1;
+            })
+    );
+
+    // Relative to byte[] array.length or MemorySegment.byteSize()
+    static final List<IntFunction<Integer>> SHUFFLE_BYTE_INDEX_GENERATORS = List.of(
+            withToString("-1", (int l) -> {
+                return -1;
+            }),
+            withToString("l", (int l) -> {
+                return l;
+            }),
+            withToString("l - 1", (int l) -> {
+                return l - 1;
+            }),
+            withToString("l + 1", (int l) -> {
+                return l + 1;
+            })
     );
 
-    static final List<IntFunction<MemorySegment>> MEMORY_SEGMENT_GENERATORS = List.of(
+    @DataProvider
+    public Object[][] shuffleIntProvider() {
+        return SHUFFLE_INT_GENERATORS.stream()
+                .map(f -> new Object[]{f})
+                .toArray(Object[][]::new);
+    }
+
+    @DataProvider
+    public Object[][] shuffleIntProviderForIOOBE() {
+        var f = SHUFFLE_INT_GENERATORS.get(0);
+        return SHUFFLE_INDEX_GENERATORS.stream()
+                .map(fi -> {
+                    return new Object[] {f, fi};
+                })
+                .toArray(Object[][]::new);
+    }
+
+    @DataProvider
+    public Object[][] shuffleIntMemorySegmentProvider() {
+        return SHUFFLE_INT_GENERATORS.stream().
+                flatMap(fa -> SHUFFLE_MEMORY_SEGMENT_GENERATORS.stream().
+                        flatMap(fb -> BYTE_ORDER_VALUES.stream().map(bo -> {
+                            return new Object[]{fa, fb, bo};
+                        }))).
+                toArray(Object[][]::new);
+    }
+
+    @DataProvider
+    public Object[][] shuffleIntByteProviderForIOOBE() {
+        var f = SHUFFLE_INT_GENERATORS.get(0);
+        return SHUFFLE_BYTE_INDEX_GENERATORS.stream().map(fi -> {
+                    return new Object[] {f, fi};
+                }).toArray(Object[][]::new);
+    }
+
+    static MemorySegment toShuffleSegment(VectorSpecies<?> vsp, int[] a, IntFunction<MemorySegment> fb) {
+        MemorySegment ms = fb.apply(a.length * 4);
+        for (int i = 0; i < a.length; i++) {
+            ms.set(ValueLayout.JAVA_INT_UNALIGNED, i * 4 , a[i]);
```

Comment:
Can we wrap the int[] in a memory segment then copy it into ms? Then we don't need to loop.